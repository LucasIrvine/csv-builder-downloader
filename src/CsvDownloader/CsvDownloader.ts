declare global {
  interface Navigator {
    msSaveBlob?: (blob: any, defaultName?: string) => boolean;
  }
}
export default class CsvDownloader {
  encodedFile: string;
  filename: string;

  constructor(encodedFile: string, filename: string) {
    this.encodedFile = encodedFile;
    this.filename = filename;
  }

  getBlob() {
    return new Blob([this.encodedFile]);
  }

  getBlobUrl(csvData: Blob) {
    if (window.URL?.createObjectURL) {
      return window.URL.createObjectURL(csvData);
    }

    return window.webkitURL.createObjectURL(csvData);
  }

  download() {
    // Proprietary to IE window.navigator
    const nav = window.navigator as any;

    if (nav.msSaveBlob) {
      nav.msSaveBlob(this.getBlob(), this.filename);

      return;
    }

    // Non-IE
    const downloadUrl = this.getBlobUrl(this.getBlob());
    const tempLink = document.createElement('a');

    tempLink.style.display = 'none';
    tempLink.href = downloadUrl;
    tempLink.setAttribute('download', this.filename);

    if (typeof tempLink.download === 'undefined') {
      tempLink.setAttribute('target', '_blank');
    }

    document.body.appendChild(tempLink);
    tempLink.click();
    document.body.removeChild(tempLink);
  }
}
