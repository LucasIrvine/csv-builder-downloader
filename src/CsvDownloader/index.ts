import CsvDownloader from './CsvDownloader';

export default CsvDownloader;
