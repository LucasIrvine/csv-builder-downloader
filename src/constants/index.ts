// Types
export type CELL_TYPE = string | number | boolean;
export type ROW_ARRAY_TYPE = (string | number | boolean)[];
export type SECTION_TYPE = { title: string; headers: ROW_ARRAY_TYPE; rows: ROW_ARRAY_TYPE[]; newLines: number };
export type CSV_BUILDER_CONFIG_TYPE = {
  encodingType?: string;
  file?: string;
  fileSuffix?: string;
  filename?: string;
  includeTimeStamp?: boolean;
  nonValueIndices?: number[];
  sanitizeRegex?: RegExp;
  sanitizeValues?: boolean;
};
// Params
export const MAX_NEW_LINES = 20;

// Regex
export const ALPHA_NUMERIC_DATE_REGEX = /[^a-zA-Z0-9:\./_\-@$ ]/g;
export const REMOVE_COMMA_REGEX = /,/g;

// Values
export const NEW_LINE = '\n';
export const COMMA = ',';
export const EMPTY_STRING = '';
export const ZERO = 0;
export const DECIMAL_ZERO = 0.0;
export const EMPTY_ARRAY = [];
export const EMPTY_OBJECT = {};
export const NA = 'N/A';

// Files
export const CSV_ENCODING_TYPE = `data:text/csv;charset=utf-8${COMMA}`;
export const DEFAULT_FILENAME = 'csv-download';
export const DEFAULT_SUFFIX = '.csv';

// Default type values
export const STRING_DEFAULT = EMPTY_STRING;
export const NUMBER_DEFAULT = ZERO;
export const ARRAY_DEFAULT = EMPTY_ARRAY;
export const OBJECT_DEFAULT = EMPTY_OBJECT;
export const NULL_DEFAULT = EMPTY_STRING;
export const DEFAULT_REGEX = ALPHA_NUMERIC_DATE_REGEX;
