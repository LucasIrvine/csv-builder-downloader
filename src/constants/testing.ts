import {
  COMMA,
  CSV_ENCODING_TYPE,
  EMPTY_STRING,
  DEFAULT_FILENAME,
  DEFAULT_SUFFIX,
  DEFAULT_REGEX,
  REMOVE_COMMA_REGEX,
} from '.';

// These are excluded from the ts build
export const CSV_BUILDER_CONFIG_SIMPLE = {
  filename: 'test-file-simple',
};
export const CHARS = `abcXYZ123,,..!!@@##$$%%^^&&**(())__++--==~~[[]]\\{{}}||;;''::""//<<>>??¡™£¢∞§¶•ªº–≠œ∑´®†¥¨ˆøπ“‘«åß∂ƒ©˙∆˚¬…æ≈ç√∫˜µ≤≥÷`;
export const DEFAULT_ALLOWED_CHARS = 'abcXYZ123..@@$$__--:://';
export const STRING_INPUT_1 = 'Is this your homework larry?';
export const STRING_INPUT_2 = 'Can you not';
export const ENCODED_STRING_INPUT = encodeURI(STRING_INPUT_2);
export const NUMERIC_INPUT = 1000000.713;
export const COMMA_INPUT = 'Well, well, well...';
export const NO_COMMA_INPUT = 'Well well well...';
export const FORMATTED_NUMERIC_INPUT = '$1,000,000.713';
export const DATE_1 = '01/04/85';
export const DATE_2 = '12-05-1983';
export const TEST_ROW = [STRING_INPUT_1, NUMERIC_INPUT, STRING_INPUT_2, FORMATTED_NUMERIC_INPUT, DATE_1, DATE_2, CHARS];
export const TEST_ROWS_ARRAY = [TEST_ROW, TEST_ROW, TEST_ROW];
export const ENCODED_FILE = `${CSV_ENCODING_TYPE}${ENCODED_STRING_INPUT}${COMMA}`;
export const MOCK_FILENAME = 'MyCsvDownload';
export const MOCK_FUNCTION = () => MOCK_FILENAME;
export const MOCK_BOOLEAN = true;
export const MOCK_OBJECT = { test: 'object' };
export const ALL_OPTIONS_DEFAULT = {
  encodingType: CSV_ENCODING_TYPE,
  file: EMPTY_STRING,
  fileSuffix: DEFAULT_SUFFIX,
  filename: DEFAULT_FILENAME,
  includeTimeStamp: true,
  nonValueIndices: [],
  sanitizeRegex: DEFAULT_REGEX,
  sanitizeValues: true,
};
export const ALL_OPTIONS_CUSTOM = {
  encodingType: 'data:text/csv;charset=utf-16',
  file: 'Already, Have, Header, Data,',
  fileSuffix: '.xlxs',
  filename: 'Not_A_Very_Good_Name',
  includeTimeStamp: false,
  nonValueIndices: [0, 1],
  sanitizeRegex: REMOVE_COMMA_REGEX,
  sanitizeValues: false,
};
