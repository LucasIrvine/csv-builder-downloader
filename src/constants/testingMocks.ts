import { COMMA, CSV_ENCODING_TYPE } from '../constants';

// These are excluded from the ts build
export const CHARS = `abcXYZ123,,..!!@@##$$%%^^&&**(())__++--==~~[[]]\\{{}}||;;''::""//<<>>??¡™£¢∞§¶•ªº–≠œ∑´®†¥¨ˆøπ“‘«åß∂ƒ©˙∆˚¬…æ≈ç√∫˜µ≤≥÷`;
export const DEFAULT_ALLOWED_CHARS = 'abcXYZ123..@@$$__--:://';
export const STRING_INPUT_1 = 'Is this your homework larry?';
export const STRING_INPUT_2 = 'Can you not';
export const ENCODED_STRING_INPUT = encodeURI(STRING_INPUT_2);
export const NUMERIC_INPUT = 1000000.713;
export const FORMATTED_NUMERIC_INPUT = '$1,000,000.713';
export const DATE_1 = '01/04/85';
export const DATE_2 = '12-05-1983';
export const TEST_ROW = [STRING_INPUT_1, NUMERIC_INPUT, STRING_INPUT_2, FORMATTED_NUMERIC_INPUT, DATE_1, DATE_2, CHARS];
export const TEST_ROWS_ARRAY = [TEST_ROW, TEST_ROW, TEST_ROW];
export const ENCODED_FILE = `${CSV_ENCODING_TYPE}${ENCODED_STRING_INPUT}${COMMA}`;
export const MOCK_FILENAME = 'MyCsvDownload';
