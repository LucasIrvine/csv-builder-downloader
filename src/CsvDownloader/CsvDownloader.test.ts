import CsvDownloader from './CsvDownloader';

import { ENCODED_FILE, MOCK_FILENAME } from '../constants/testing';

describe('CsvDownloader', () => {
  let csvDownloader: CsvDownloader;

  beforeEach(() => {
    csvDownloader = new CsvDownloader(ENCODED_FILE, MOCK_FILENAME);
  });

  describe('.getBlob', () => {
    it('should return an instance of a Blob', () => {
      const csvBlob = csvDownloader.getBlob();

      expect(csvBlob instanceof Blob).toEqual(true);
    });
  });

  describe('.getBlobUrl', () => {
    it('should call window.URL.createObjectURL if supported', () => {
      global.URL.createObjectURL = jest.fn();
      const testBlob = csvDownloader.getBlob();

      csvDownloader.getBlobUrl(testBlob);

      expect(global.URL.createObjectURL).toHaveBeenCalledTimes(1);
    });

    it('should fallback to webkitURL if window.URL.createObjectURL is not supported', () => {
      // @ts-ignore: passing in the wrong type
      global.URL.createObjectURL = null;

      global.webkitURL.createObjectURL = jest.fn();
      const testBlob = csvDownloader.getBlob();

      csvDownloader.getBlobUrl(testBlob);

      expect(window.webkitURL.createObjectURL).toHaveBeenCalledTimes(1);
    });

    it('should return the same output given the same input', () => {
      const testBlob = csvDownloader.getBlob();

      expect(csvDownloader.getBlobUrl(testBlob)).toMatchSnapshot();
    });
  });

  describe('.download', () => {
    it('should use msSaveBlob to download if IE browser and skip the link building', () => {
      global.navigator.msSaveBlob = jest.fn();
      global.document.createElement = jest.fn();

      csvDownloader.download();

      expect(global.navigator.msSaveBlob).toHaveBeenCalledTimes(1);
      expect(global.document.createElement).toHaveBeenCalledTimes(0);
    });
  });
});
