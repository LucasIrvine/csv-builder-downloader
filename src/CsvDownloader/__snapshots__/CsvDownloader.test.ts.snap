// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvDownloader .getBlobUrl should return the same output given the same input 1`] = `undefined`;
